File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_4.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var postorderTraversal = function(root) 
{
    const values = [];
    if (!root) return values;

    const tree = [root];

    while (tree.length > 0) {
        const root = tree.pop();

        values.unshift(root.val);

        if (root.left)
            tree.push(root.left);

        if (root.right)
            tree.push(root.right);
    }

    return values;    
};

File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 class Solution {
    public List<Integer> postorderTraversal(TreeNode root) 
    {
        List<Integer> values = new ArrayList<>();
        
        if (root == null) return values;

        List<TreeNode> tree = new ArrayList<>();
        tree.add(root);

        while (!tree.isEmpty()) {
            root = tree.get(tree.size() - 1);
            tree.remove(tree.size() - 1);

            values.add(0, root.val);

            if (root.left != null)
                tree.add(root.left);

            if (root.right != null)
                tree.add(root.right);
        }

        return values;
    }
}